### PR TITLE
Fix process failure protocol

### PR DESCRIPTION
**Pull request by vsreekanti**
This PR is step 1 of fixing the problems described in #2 by addressing process failure. 

When a server process fails, Kubernetes will immediately restart it on the same machine where it was running before, meaning it has the same IP address. We use this fact to avoid updating the hash ring and regossip the data. Changes in this PR include:

* Pods now automatically restart when the process fails.
* The `kops-pod` responds to requests for restart count of a pod.
* The hash ring API accounts for a restart count per node -- when a node is restarting, the hash ring returns `true` on insertion, even if the hash ring size hasn't changed.

### CODE REVIEW

File: kvs/include/hash_ring.hpp [hunk 1]
```diff
@@ -32,30 +32,47 @@ class HashRing : public ConsistentHashMap<ServerThread, H> {
     return unique_servers;
   }
 
-  bool insert(Address public_ip, Address private_ip, unsigned tid) {
+  bool insert(Address public_ip, Address private_ip, int join_count,
+              unsigned tid) {
     bool succeed;
```
Comment (cw75) [inline kvs/include/hash_ring.hpp]: variable not used

File: kvs/src/kvs/node_join_handler.cpp [hunk 2]
```diff
@@ -86,7 +90,17 @@ void node_join_handler(
...
+          // node:
+          // 1) if the node is a new node and I am no longer responsible for
+          // the key
+          // 2) if the node is rejoining the cluster
+          // NOTE: This is currently inefficient because every server will
+          // gossip the key currently -- we might be able to hack around the
+          // has ring to do it more efficiently, but I'm leaving this here for
+          // now
+          if ((join_count == 0 && threads.find(wt) == threads.end()) ||
+              join_count > 0) {
```
Comment (cw75) [inline kvs/src/kvs/node_join_handler.cpp]: Seems like currently all threads gossip _all_ keys to the new node if its `join_count` > 0, but shouldn't we first check if the rejoining node/threads are responsible for the key to be more efficient?
Also, we probably shouldn't put the key into `join_remove_set` for the rejoining case, as the keys in this set will eventually be deleted, which we may not want.

File: kvs/src/kvs/user_request_handler.cpp [hunk 3]
```diff
@@ -78,6 +78,11 @@ void user_request_handler(
         KeyTuple* tp = response.add_tuples();
         tp->set_key(key);
 
+        if (!is_metadata(key)) {
```
Comment (cw75) [inline kvs/src/kvs/user_request_handler.cpp]: Do we want to always log the keys being processed? Or is this for debugging only?

Comment (vsreekanti) [reply kvs/src/kvs/user_request_handler.cpp]: I intended for this to be general logging, but I don't feel strongly about it if you think it should be removed.

**Pull request merged**